CIPs/0017.md

```diff
@@ -0,0 +1,101 @@
+```
+
+The `les` client must then be modified to send a `GetCertifiedEpochsValidatorSet` message during `lightest` sync. This will likely replace the function `getEpochOrNormalHeaders` within `eth/downloader`.
+
+Upon receiving the `CertifiedEpochsValidatorSet` response message, the light client must then verify headers and proofs provided. Proofs will be verified via a call into `celo-bls-go`'s `verify` fn.
+
+The verify function requires information on an epoch's validator set, so this will likely require some access changes (private -> public) and exposing more data structures to allow the validator set to be queried and modified upon proof verification.
+
+### Upgradeability
+In order to support an upgradeable snark circuit, we utilize a `versionNumber` parameter with every proof. This corresponds to a version of the snark circuit and thus a corresponding verification key. The clients are expected to be shipped with all available verification keys, and can be upgraded to receive added verification keys.
+
+For any uses of the snark's `verify` fn in the above protocols, a client is expected to:
+* Fetch the verification key corresponding to the proof's `versionNumber` from its storage
+* The client then uses that verification key to call `verify` on the proof.
+
+### Bad proofs
+Bad proofs should be prevented from propagated throughout the network via the `verify` check that each node performs. In the case that any participant receives a proof that does not pass verification for its inputs, that participant is expected to drop the sender. 
+
+## Alternative Solutions
+
+Alternatively, `GetPlumoProofMsg` could utilize a bloom filter instead, allowing for a constant sized message. However, bloom filters allow for false positives: a false positive here would indicate to a peer that the sender has a proof when the sender does not. This could prevent certain proofs from ever getting to some nodes. For example, say node A is on the network and has received a certain amount of proofs via traditional propagation. Node A then goes down for a short period, and within that period proof P is added and propagated. Node A then comes back online, and sends a `GetPlumoProofsMsg` to request any proofs it may have missed, and sends along the corresponding bloom filter for the proofs it holds. If proof P happens to be a false positive for the set of proofs node A already has, then all of node A's peers will simply return nothing. Thus node A will never receive proof P. This can then propagate similarly through the network either through a similar path as node A, or for any nodes where all of their peers have already "lost" proof P.
+
+`GetPlumoProofMsg` could also request *all* plumo proofs from the receiver, leaving the responsibility of filtering seen proofs with the caller. This is a simpler implementation, but is less scalable as we increase the amount of stored proofs. 
```

Comment:
qq -- as a naive reader, I'd be curious, do all proofs have to be stored in perpetuity? Eg. Proof A maps from epoch 0 to epoch 10, proof B from epoch 11-20, incrementing up to epoch 1000. Do we store all 100 proofs?

Wondering why a prover wouldn't aggregate this into one proof from epoch 0-1000?

Reply:
Not naive at all - this is something we've been mulling over as well.
The circuit that generates these proofs has a "hard-coded" limit of how many epochs a proof can span (I believe right now we're targeting 120*), so while we can't roll up to 1000 like in your example, we could roll up to this limit. 

That said, while we can reduce redundant proofs like you say, currently I'm thinking we shouldn't. The reasons being that `lightest` clients may still ask about blocks within epochs between boundaries, so removing intermediary proofs could hurt the performance there. The other reason is that proofs aren't very large and there likely won't be too many, so the tradeoffs seem worth it here.

* we may be able to change this limit dynamically, but that's not super relevant here